cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -261,58 +331,75 @@ protected void rebalanceForBroker(Broker broker,
+                                          ClusterModel clusterModel,
+                                          Set<Goal> optimizedGoals,
+                                          ActionType actionType,
+                                          Set<String> excludedTopics) {
+
+    if (!clusterModel.newBrokers().isEmpty() && !broker.isNew()) {
+      // We have new brokers and the current broker is not a new broker.
+      return true;
+    }
+
+    PriorityQueue<CandidateBroker> candidateBrokerPQ = new PriorityQueue<>();
+    // Sort the replicas initially to avoid sorting it every time.
+
+    double clusterUtilization = clusterModel.load().expectedUtilizationFor(resource()) / clusterModel.capacityFor(resource());
+    for (Broker candidate : clusterModel.healthyBrokers()) {
+      // Get candidate replicas on candidate broker to try moving load from -- sorted in the order of trial (descending load).
+      if (utilizationPercentage(candidate) > clusterUtilization) {
+        SortedSet<Replica> replicasToMoveIn = sortedCandidateReplicas(candidate, excludedTopics, 0, true);
+        CandidateBroker candidateBroker = new CandidateBroker(candidate, replicasToMoveIn, false);
+        candidateBrokerPQ.add(candidateBroker);
       }
-    });
+    }
 
     // Stop when all the replicas are leaders for leader movement or there is no replicas can be moved in anymore
     // for replica movement.
-    while (!eligibleBrokers.isEmpty() && (actionType == REPLICA_MOVEMENT ||
+    while (!candidateBrokerPQ.isEmpty() && (actionType == REPLICA_MOVEMENT ||
         (actionType == LEADERSHIP_MOVEMENT && broker.leaderReplicas().size() != broker.replicas().size()))) {
-      Broker sourceBroker = eligibleBrokers.poll();
-      for (Replica replica : sourceBroker.sortedReplicas(resource())) {
+      CandidateBroker cb = candidateBrokerPQ.poll();
+      SortedSet<Replica> candidateReplicasToReceive = cb.replicas();
+
+      Set<Replica> receivedReplicas = new HashSet<>();
+      for (Replica replica : candidateReplicasToReceive) {
         if (shouldExclude(replica, excludedTopics)) {
           continue;
         }
-        // It does not make sense to move a replica without utilization from a live broker.
-        if (replica.load().expectedUtilizationFor(resource()) == 0.0 && broker.isAlive()) {
-          break;
-        }
+
         Broker b = maybeApplyBalancingAction(clusterModel, replica, Collections.singletonList(broker), actionType, optimizedGoals);
         // Only need to check status if the action is taken. This will also handle the case that the source broker
         // has nothing to move in. In that case we will never reenqueue that source broker.
         if (b != null) {
-          if (isLoadAboveBalanceLowerLimit(clusterModel, broker)) {
+          if (isLoadAboveBalanceLowerLimit(broker)) {
             return false;
           }
+          // Identify the replica that the source broker has received.
+          receivedReplicas.add(replica);
```

Comment:
Why do we need to keep a set for this the received replicas? Can we remove the replicas immediately?

Reply:
The loop is iterating over `candidateReplicasToReceive`. Removing replicas from this collection immediately would cause `ConcurrentModificationException`.